### PR TITLE
PSBTv2 (BIP-370) Support

### PR DESCRIPTION
**Pull request by notTanveer**
This PR improves PSBT (Partially Signed Bitcoin Transaction) v2 support, especially strict validation and compatibility with BIP 370 test vectors. It also adds a comprehensive set of tests for PSBTv2 edge cases and locktime determination.

---

### 1. **PSBTv2 Compliance and Parsing Improvements**
- Updated method signatures in `src/embit/liquid/pset.py` to allow passing the `version` parameter down to `read_value` and its parent implementations. This ensures correct handling of versioned PSBT fields.
- Adjusted calls to `super().read_value` to pass along the `version` argument, improving extensibility and correctness for PSBTv2.

### 2. **PSBT View Updates**
- Modified `src/embit/psbtview.py`:
  - The `input()` and `output()` methods now pass the `version` parameter when reading from the stream, allowing PSBT-in and out scopes to parse versioned fields correctly.

### 3. **Comprehensive PSBTv2 Test Vectors**
- Added a new test file: `tests/tests/test_psbtV2.py` containing:
  - Test vectors based on [BIP 370 PSBTv2](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#user-content-Test_Vectors).
  - Tests for invalid PSBTv0/PSBTv2 edge cases (e.g., wrong/missing fields, invalid locktimes).
  - Tests for valid PSBTv2 minimal and full-featured transactions.
  - Tests for locktime determination logic in PSBTv2, including edge cases with multiple locktimes (height and time).

---

## Motivation

- Ensures full BIP 370 compliance, increasing interoperability and correctness for PSBTv2 parsing and validation.
- Catches subtle bugs and regressions through comprehensive test coverage.
- Lays the groundwork for future PSBTv2 features and broader Elements/Liquid support.

---

## Testing

- Run `pytest tests/tests/test_psbtV2.py` to verify all BIP 370 test vectors and locktime cases.
- Built from the referenced PR and generated various PSBTs, then imported them into `embit` to confirm correct parsing — everything worked as expected.
- Would love for others to build Bitcoin Core + CLI from [PR #21283](https://github.com/bitcoin/bitcoin/pull/21283), try it out, and share your observations here.

---

## References

- [BIP 370: PSBTv2](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki)
- [Bitcoin Core PR #21283](https://github.com/bitcoin/bitcoin/pull/21283)

### CODE REVIEW

File: src/embit/psbt.py [hunk 1]
```diff
@@ -345,13 +357,48 @@ def read_value(self, stream, k):
...
+            if version != 2:
+                raise PSBTError(
+                    "PSBT_IN_REQUIRED_HEIGHT_LOCKTIME not allowed in PSBTv0"
+                )
+            if len(k) != 1:
+                raise PSBTError("Invalid required height locktime key")
+            if self.required_height_locktime is not None:
+                raise PSBTError("Duplicated required height locktime")
+            locktime = int.from_bytes(v, "little")
+            if locktime >= 500000000 or locktime == 0:
```
[the remark this answers is not in the document]
Comment (jdlcdl) [reply src/embit/psbt.py:398]: I don't think `int.from_bytes("little")` can give us a negative integer.

Im always a fan of `if not too_few < a_value < too_many:` but it's just style.

In this case, 500 million is such an important and dizzying number -- also used above, that it might deserve to be a constant like MIN_LOCKTIME_TIMESTAMP.

**Review by odudex — COMMENTED** [review]
Please rebase with master (we plan to stop using develop branch). There shouldn't be conflicts rebasing it locally with git.

Later, there are a few things to take a look at:

1. PSBTView does not fully enforce or apply PSBTv2 rules. It can accept a v2 PSBT with PSBT_GLOBAL_UNSIGNED_TX if that key appears before PSBT_GLOBAL_VERSION, and PSBTView.locktime uses only fallback locktime instead of deriving locktime from per-input requirements. Sighashes computed through PSBTView can therefore commit to the wrong locktime.

2. add_input() / add_output() only check modifiable flags. BIP370 Constructor rules also require validating required fields, locktime compatibility, preserving existing signed locktime, and handling SIGHASH_SINGLE input/output pairing.

3. Several PSBTv2 fixed-width input fields lack exact length validation: previous txid, output index, sequence, required time locktime, and required height locktime. Malformed values can be accepted and later serialize incorrectly or fail unexpectedly.

4. compress mode can lose its OOM-protection benefit for PSBTv2 non-witness UTXOs when PSBT_IN_NON_WITNESS_UTXO appears before PSBT_IN_PREVIOUS_TXID / PSBT_IN_OUTPUT_INDEX in the input map. Since key order is not guaranteed, this should be pre-scanned, deferred, or documented.

5. Signedness should be made consistent with BIP370: PSBT_OUT_AMOUNT is read as signed but written unsigned, and PSBT_GLOBAL_TX_VERSION is specified as signed int32 but read/written unsigned.

6. Explicit PSBTv0 version handling is too strict. BIP174 allows v0 PSBTs to either omit `PSBT_GLOBAL_VERSION` or set it to `0`, but this parser rejects version `0`.

7. PSBTv2 global input/output counts are used to allocate placeholder scope objects before parsing actual scopes. Since those counts are attacker-controlled and the placeholders are later discarded, this creates an avoidable memory-exhaustion risk. Store counts only and allocate scopes as they are parsed.

Recommended test additions: PSBTv2 parse/serialize roundtrips, PSBTView locktime/sighash behavior, constructor API behavior, signer TX_MODIFIABLE updates, malformed fixed-width field lengths, and the PSET v2 output validation path.

File: src/embit/psbt.py [hunk 2]
```diff
@@ -1057,13 +1420,116 @@ def sign_with(self, root, sighash=SIGHASH.DEFAULT) -> int:
...
+
+        is_anyonecanpay = bool(inp_sighash & SIGHASH.ANYONECANPAY)
+        sighash_type = inp_sighash & 0x1F
+
+        if self.tx_modifiable_flags is None:
+            self.tx_modifiable_flags = 0
+
+        if not is_anyonecanpay:
+            self.tx_modifiable_flags &= ~TxModifiable.INPUTS
+        if sighash_type != SIGHASH.NONE:
```
[the remark this answers is not in the document]
Comment (notTanveer) [reply src/embit/psbt.py:1473]: read the [spec](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#signer) properly

> If the Signer added a signature that does not use SIGHASH_NONE, the Outputs Modifiable flag must be set to False.

**Comment by odudex** [comment]
I sent you a PR with final nits and adaptations for older MicroPython versions, like the one we run in Krux.
I did very basic tests, and with the commits I sent you, is works with Krux.

Please evaluate the commits and merge them if you agree with changes. Let me know otherwise

**Review by odudex — DISMISSED** [review]
tACK

I did basic testing with Krux, though nothing specific to the newly added PSBTv2 features. Basic functionality worked after replacing a few syntax constructs with simpler equivalents compatible with Krux’s older MicroPython version.

**Review by qlrd — COMMENTED** [review]
nit, not needed

**Review by qlrd — COMMENTED** [review]
jut nit, not necessary

File: src/embit/psbt.py [hunk 3]
```diff
@@ -245,7 +266,7 @@ def verify(self, ignore_missing=False):
             raise PSBTError("Missing non_witness_utxo")
         return False
 
-    def read_value(self, stream, k):
+    def read_value(self, stream, k, version=None):
```
Comment (qlrd) [inline src/embit/psbt.py:269]: Maybe

```suggestion
    def read_value(self, stream, k, version=None, *args, **kwargs):
```

could supress some warnings? (not necessary although, since the `version` will be passed to `**kwargs` in PSBT, just a CI stuff)

**Review by qlrd — COMMENTED** [review]
tACK, just nit:

File: src/embit/psbt.py (hunk 3 above)
Comment (odudex) [reply src/embit/psbt.py:269]: What warning?

Comment (qlrd) [reply src/embit/psbt.py:269]: i saw one by CodeQL like this:

```bash
Mismatch between signature and use of an overridden method
  Check notice: Mismatch between signature and use of an overridden method
  Overridden method signature does not match [call](1), where it is passed an argument named 'version'. Overriding method [method
  InputScope.read_value](2) matches the call.
  Overridden method signature does not match [call](3), where it is passed an argument named 'version'. Overriding method [method
  OutputScope.read_value](4) matches the call.
```

when `read_value` is called on tests, but idk why disapeared

Comment (odudex) [reply src/embit/psbt.py:269]: Good catch, maybe it's simpler to change parent definition:
```diff
# Line 101
-    def read_value(self, stream, key, *args, **kwargs):
+    def read_value(self, stream, key, version=None):
```

File: src/embit/psbtview.py [hunk 4]
```diff
@@ -411,7 +413,7 @@ def vout(self, i, compress=None):
 
         self.seek_to_scope(self.num_inputs + i)
         v = self.get_value(b"\x03", from_current=True)
-        value = int.from_bytes(v, "little")
+        value = _signed_from_bytes(v)
```
Comment (notTanveer) [inline src/embit/psbtview.py:416]: we didn't have `signed=True` here. Any reason why we replaced this with `_signed_from_bytes` in this case? @odudex

Comment (odudex) [reply src/embit/psbtview.py:416]: For normal Bitcoin amounts this has no practical behavioral difference, since valid amounts are non-negative. So it is not critical, but it keeps PSBTView consistent with BIP370 and with the regular parser.
It is not needed for the MicroPython fix specifically, it's just for consistency.
PSBTView.vout() is reading PSBT_OUT_AMOUNT, and BIP370 defines that field as a signed 64-bit little-endian integer. The full OutputScope parser already decodes it with _signed_from_bytes, so PSBTView should decode the same field the same way.
